samcli/hook_packages/terraform/hooks/prepare.py

```diff
@@ -301,6 +301,124 @@ def _validate_referenced_resource_matches_sam_metadata_type(
+        as json string
+    src_code_attribute_name: str
+        the lambda function source code or docker context to be used to raise the correct exception
+
+    Returns
+    -------
+    str
+        The lambda function source code or docker context paths
+    """
+    LOG.info(
+        "Extract the %s from the sam metadata resource %s from property %s",
+        src_code_attribute_name,
+        sam_metadata_resource_address,
+        src_code_property_name,
+    )
+    source_code = sam_metadata_attributes.get(src_code_property_name)
+    source_code_property = sam_metadata_attributes.get(property_path_property_name)
+    LOG.debug(
+        "The found %s value is %s and property value is %s", src_code_attribute_name, source_code, source_code_property
+    )
+    if not source_code:
+        raise InvalidSamMetadataPropertiesException(
+            f"The sam metadata resource {sam_metadata_resource_address} should contain the lambda function "
+            f"{src_code_attribute_name} in property {src_code_property_name}"
+        )
+    if isinstance(source_code, str):
+        try:
+            LOG.debug("Try to decode the %s value in case if it is a encoded JSON string.", src_code_attribute_name)
+            source_code = json.loads(source_code)
+            LOG.debug("The decoded value of the %s value is %s", src_code_attribute_name, source_code)
+        except Exception:
+            LOG.debug("Source code value could not be parsed as a JSON object. Handle it as normal string value")
+
+    if isinstance(source_code, dict):
+        LOG.debug(
+            "Process the extracted %s as JSON object using the property %s",
+            src_code_attribute_name,
+            source_code_property,
+        )
+        if not source_code_property:
+            raise InvalidSamMetadataPropertiesException(
+                f"The sam metadata resource {sam_metadata_resource_address} should contain the lambda function "
+                f"{src_code_attribute_name} property in property {property_path_property_name} as the "
+                f"{src_code_property_name} value is an object"
+            )
+        cfn_source_code_path = source_code.get(source_code_property)
+        if not cfn_source_code_path:
+            LOG.error(
+                "The property %s does not exist in the extracted %s JSON object %s",
+                source_code_property,
+                src_code_attribute_name,
+                source_code,
+            )
+            raise InvalidSamMetadataPropertiesException(
+                f"The sam metadata resource {sam_metadata_resource_address} should contain a valid lambda function "
+                f"{src_code_attribute_name} property in property {property_path_property_name} as the "
+                f"{src_code_property_name} value is an object"
+            )
+    elif isinstance(source_code, list):
+        # SAM CLI does not process multiple paths, so we will handle only the first value in this list
```

Comment:
should we say this in an info log to inform users?

Reply:
hmmm .. not sure, I was thinking of documenting this behaviour, but we can add logs as well if the list contains more than 1 item.

Reply:
Done